contracts/RewardPool.sol

```diff
@@ -256,6 +289,179 @@ contract RewardPool is AccessControl, ReentrancyGuard {
+   * @param periodStart Start of the reward period (unix timestamp)
+   * @param periodEndExclusive End of the reward period (unix timestamp)
+   * @param kpiFunctionId Bytes32 identifier of the kpi function (e.g., github commit hash)
+   * @notice Allowed only for address with DEFAULT_ADMIN_ROLE
+   */
+  function updatePeriodKpis(
+    IRewardFunction.Kpi[] calldata kpis,
+    uint48 periodStart,
+    uint48 periodEndExclusive,
+    bytes32 kpiFunctionId
+  ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+    bytes32 periodId = _getPeriodId(periodStart, periodEndExclusive);
+
+    PeriodData storage period = _periods[periodId];
+
+    if (period.processedAt != 0) revert PeriodAlreadyProcessed(periodId);
+
+    for (uint256 i = 0; i < kpis.length; i++) {
+      if (kpis[i].referrerAddress == address(0))
+        revert ZeroAddressNotAllowed(i);
+
+      period.users.add(kpis[i].referrerAddress);
+      period.kpis[kpis[i].referrerAddress] = kpis[i].kpi;
+
+      emit KpiUpdated(
+        kpis[i].referrerAddress,
+        periodId,
+        kpis[i].kpi,
+        periodStart,
+        periodEndExclusive,
+        kpiFunctionId
+      );
+    }
+  }
+
+  /**
+   * @dev Allows the owner to process a reward period and distribute rewards
+   * @param periodStart Start of the reward period (unix timestamp)
+   * @param periodEndExclusive End of the reward period (unix timestamp)
+   * @param totalRewardAmount Total amount of rewards to distribute
+   * @notice Allowed only for address with DEFAULT_ADMIN_ROLE
+   */
+  function processPeriod(
+    uint48 periodStart,
+    uint48 periodEndExclusive,
+    uint256 totalRewardAmount
+  ) external onlyRole(DEFAULT_ADMIN_ROLE) {
+    bytes32 periodId = _getPeriodId(periodStart, periodEndExclusive);
+
+    PeriodData storage period = _periods[periodId];
+
+    if (period.processedAt != 0) revert PeriodAlreadyProcessed(periodId);
+
+    if (period.users.length() == 0)
+      revert PeriodInvalid(periodStart, periodEndExclusive);
+
+    IRewardFunction.Kpi[] memory kpis = new IRewardFunction.Kpi[](
+      period.users.length()
+    );
+
+    for (uint256 i = 0; i < period.users.length(); i++) {
+      address user = period.users.at(i);
+      uint256 kpi = period.kpis[user];
+
+      kpis[i] = IRewardFunction.Kpi({referrerAddress: user, kpi: kpi});
+    }
+
+    IRewardFunction.Reward[] memory rewards = IRewardFunction(
+      rewardFunctionAddress
+    ).calculateReward(kpis, totalRewardAmount);
+
+    uint256[] memory rewardFunctionArgs = new uint256[](2);
+    rewardFunctionArgs[0] = periodStart;
+    rewardFunctionArgs[1] = periodEndExclusive;
+
+    uint256 totalIssuedRewardAmount = 0;
+    uint256 numUsersRewarded = 0;
+
+    for (uint32 i = 0; i < rewards.length; i++) {
+      if (rewards[i].reward == 0) continue;
+
+      RewardData memory rewardData = RewardData({
+        user: rewards[i].referrerAddress,
+        amount: rewards[i].reward,
+        idempotencyKey: keccak256(
+          abi.encode(
```

Comment:
I'm not sure it really matters (since any campaign that uses this version of the contract will not be generating idempotency keys off-chain), but the [idempotency key generation in the off-chain scripts](https://github.com/divvi-xyz/divvi-protocol/blob/8ece6ef91f7c9339da73f46cf646b5c8c7569d69/scripts/utils/createSafeTransactionsBatch.ts#L69) differs slightly from the format used here -- I wonder if we should update the off-chain idempotency key generation for consistency's sake. Not a big deal though.

Reply:
Probably not a big deal since we ideally should never call `addRewards` if we are using kpis. We can update our off-chain script but one side effect is that active campaigns (tether) has already used the existing format, so maybe fine as is?

Reply:
yeah, I think fine as-is!